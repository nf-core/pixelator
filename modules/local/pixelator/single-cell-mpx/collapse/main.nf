process PIXELATOR_COLLAPSE {
    tag "${meta.id}"
    label 'process_medium'
    label 'process_long'

    // TODO: Add conda back
    // conda "${moduleDir}/environment.yml"
    container "${workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container
        ? 'quay.io/pixelgen-technologies/pixelator:0.21.2'
        : 'quay.io/pixelgen-technologies/pixelator:0.21.2'}"

    input:
    tuple val(meta), path(reads), path(panel_file), val(panel)

    output:
    tuple val(meta), path("collapse/*.collapsed.parquet"), emit: collapsed
    tuple val(meta), path("collapse/*.report.json"),       emit: report_json
    tuple val(meta), path("collapse/*.meta.json"),         emit: metadata
    tuple val(meta), path("*pixelator-collapse.log"),      emit: log

    path "versions.yml",                                   emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    assert meta.design != null

    def prefix = task.ext.prefix ?: "${meta.id}"
    def args = task.ext.args ?: ''
    def reads_arg = reads.join(' ')
    def panel_opt = (panel
        ? "--panel ${panel}"
        : panel_file
            ? "--panel ${panel_file}"
            : "")

    """
    pixelator \\
        --cores ${task.cpus} \\
        --log-file ${prefix}.pixelator-collapse.log \\
        --verbose \\
        single-cell-mpx \\
        collapse \\
        --output . \\
        --design ${meta.design} \\
        ${panel_opt} \\
        ${args} \\
        ${reads_arg}

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        pixelator: \$(echo \$(pixelator --version 2>/dev/null) | sed 's/pixelator, version //g' )
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"

    """
    mkdir collapse
    touch "${prefix}.pixelator-collapse.log"
    touch "collapse/${prefix}.collapsed.parquet"
    touch "collapse/${prefix}.report.json"
    touch "collapse/${prefix}.meta.json"

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        pixelator: \$(echo \$(pixelator --version 2>/dev/null) | sed 's/pixelator, version //g' )
    END_VERSIONS
    """
}
