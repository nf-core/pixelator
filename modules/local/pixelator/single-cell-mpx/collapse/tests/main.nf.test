nextflow_process {
    name "Test Process PIXELATOR_COLLAPSE"
    script "../main.nf"
    process "PIXELATOR_COLLAPSE"
    tag "modules"
    tag "pixelator"
    tag "pixelator/mpx"
    tag "pixelator/single_cell_mpx_collapse"


    test("Test MPX collapse - stub") {

        options "-stub"

        when {
            process {
                """
                demuxed_input = [
                    file(params.pipelines_testdata_base_path + 'testdata/mpx/modules/demux/sample01_1k_pbmcs_scsp_v1_immunology1.processed-AAAGCAAC.fastq.gz', checkIfExists: true),
                    file(params.pipelines_testdata_base_path + 'testdata/mpx/modules/demux/sample01_1k_pbmcs_scsp_v1_immunology1.processed-AACAAGAC.fastq.gz', checkIfExists: true),
                    file(params.pipelines_testdata_base_path + 'testdata/mpx/modules/demux/sample01_1k_pbmcs_scsp_v1_immunology1.processed-AACCTTCC.fastq.gz', checkIfExists: true)
                ]

                input[0] = [
                    [ id:'sample01_1k_pbmcs_scsp_v1_immunology1', design:'D21', panel: "human-sc-immunology-spatial-proteomics-1", technology:'mpx' ],
                    demuxed_input,
                    [],
                    "human-sc-immunology-spatial-proteomics-1"
                ]
                input.dump()
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }


    test("Test MPX collapse - - SCSP v1 | Immunology-I") {
        when {
            params {
                save_all = true
            }
            process {
                """
                demuxed_input = [
                    file(params.pipelines_testdata_base_path + 'testdata/mpx/modules/demux/sample01_1k_pbmcs_scsp_v1_immunology1.processed-AAAGCAAC.fastq.gz', checkIfExists: true),
                    file(params.pipelines_testdata_base_path + 'testdata/mpx/modules/demux/sample01_1k_pbmcs_scsp_v1_immunology1.processed-AACAAGAC.fastq.gz', checkIfExists: true),
                    file(params.pipelines_testdata_base_path + 'testdata/mpx/modules/demux/sample01_1k_pbmcs_scsp_v1_immunology1.processed-AACCTTCC.fastq.gz', checkIfExists: true),
                    file(params.pipelines_testdata_base_path + 'testdata/mpx/modules/demux/sample01_1k_pbmcs_scsp_v1_immunology1.processed-AACGGCTA.fastq.gz', checkIfExists: true),
                    file(params.pipelines_testdata_base_path + 'testdata/mpx/modules/demux/sample01_1k_pbmcs_scsp_v1_immunology1.processed-AAGCATAG.fastq.gz', checkIfExists: true),
                ]

                input[0] = [
                    [ id:'sample01_1k_pbmcs_scsp_v1_immunology1', design:'D21', panel: "human-sc-immunology-spatial-proteomics-1", technology:'mpx' ],
                    demuxed_input,
                    [],
                    "human-sc-immunology-spatial-proteomics-1"
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                        process.out.collapsed,
                        process.out.report_json,
                        process.out.metadata
                    ).match()
                },
                {
                    with (process.out.log) {
                        assert path(get(0).get(1)).readLines().last().contains("Finished pixelator collapse")
                    }
                },
            )
        }
    }
}
