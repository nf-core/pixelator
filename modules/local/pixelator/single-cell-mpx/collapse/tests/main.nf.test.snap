{
    "Test MPX collapse - - SCSP v1 | Immunology-I": {
        "content": [
            [
                [
                    {
                        "id": "sample01_1k_pbmcs_scsp_v1_immunology1",
                        "design": "D21",
                        "panel": "human-sc-immunology-spatial-proteomics-1",
                        "technology": "mpx"
                    },
                    "sample01_1k_pbmcs_scsp_v1_immunology1.collapsed.parquet:md5,309a0b8c40bfb3e7cc6f05bd9a94b9e9"
                ]
            ],
            [
                [
                    {
                        "id": "sample01_1k_pbmcs_scsp_v1_immunology1",
                        "design": "D21",
                        "panel": "human-sc-immunology-spatial-proteomics-1",
                        "technology": "mpx"
                    },
                    "sample01_1k_pbmcs_scsp_v1_immunology1.report.json:md5,dd843f96bca3802fa558afefabf1864f"
                ]
            ],
            [
                [
                    {
                        "id": "sample01_1k_pbmcs_scsp_v1_immunology1",
                        "design": "D21",
                        "panel": "human-sc-immunology-spatial-proteomics-1",
                        "technology": "mpx"
                    },
                    "sample01_1k_pbmcs_scsp_v1_immunology1.meta.json:md5,69a3436261bc4e09d841b7be641aafbd"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.6"
        },
        "timestamp": "2025-05-06T15:11:44.266641701"
    },
    "Test MPX collapse - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "sample01_1k_pbmcs_scsp_v1_immunology1",
                            "design": "D21",
                            "panel": "human-sc-immunology-spatial-proteomics-1",
                            "technology": "mpx"
                        },
                        "sample01_1k_pbmcs_scsp_v1_immunology1.collapsed.parquet:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "sample01_1k_pbmcs_scsp_v1_immunology1",
                            "design": "D21",
                            "panel": "human-sc-immunology-spatial-proteomics-1",
                            "technology": "mpx"
                        },
                        "sample01_1k_pbmcs_scsp_v1_immunology1.report.json:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "sample01_1k_pbmcs_scsp_v1_immunology1",
                            "design": "D21",
                            "panel": "human-sc-immunology-spatial-proteomics-1",
                            "technology": "mpx"
                        },
                        "sample01_1k_pbmcs_scsp_v1_immunology1.meta.json:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "sample01_1k_pbmcs_scsp_v1_immunology1",
                            "design": "D21",
                            "panel": "human-sc-immunology-spatial-proteomics-1",
                            "technology": "mpx"
                        },
                        "sample01_1k_pbmcs_scsp_v1_immunology1.pixelator-collapse.log:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "4": [
                    "versions.yml:md5,85c7963b2a468838af56848f81fdfc9a"
                ],
                "collapsed": [
                    [
                        {
                            "id": "sample01_1k_pbmcs_scsp_v1_immunology1",
                            "design": "D21",
                            "panel": "human-sc-immunology-spatial-proteomics-1",
                            "technology": "mpx"
                        },
                        "sample01_1k_pbmcs_scsp_v1_immunology1.collapsed.parquet:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "log": [
                    [
                        {
                            "id": "sample01_1k_pbmcs_scsp_v1_immunology1",
                            "design": "D21",
                            "panel": "human-sc-immunology-spatial-proteomics-1",
                            "technology": "mpx"
                        },
                        "sample01_1k_pbmcs_scsp_v1_immunology1.pixelator-collapse.log:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "metadata": [
                    [
                        {
                            "id": "sample01_1k_pbmcs_scsp_v1_immunology1",
                            "design": "D21",
                            "panel": "human-sc-immunology-spatial-proteomics-1",
                            "technology": "mpx"
                        },
                        "sample01_1k_pbmcs_scsp_v1_immunology1.meta.json:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "report_json": [
                    [
                        {
                            "id": "sample01_1k_pbmcs_scsp_v1_immunology1",
                            "design": "D21",
                            "panel": "human-sc-immunology-spatial-proteomics-1",
                            "technology": "mpx"
                        },
                        "sample01_1k_pbmcs_scsp_v1_immunology1.report.json:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,85c7963b2a468838af56848f81fdfc9a"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.2"
        },
        "timestamp": "2025-06-26T11:51:04.702563951"
    }
}